### ISSUE TITLE
AGL's Elligator Code Might be Flawed

### ISSUE THREAD

**Issue by Daeinar** [dedis/kyber#117]
@isislovecruft and @hdevalence notified me that AGL's Elligator code might be flawed and thus should not be used. We should follow up with them on the exact details. A good documentation for Elligator that could be used for a reimplementation is https://whispersystems.org/docs/specifications/xeddsa/#hashing-to-a-point-with-elligator-2.

**Comment by Liamsi** [comment]
Interesting! Could you provide me with details what the exact flaw is (later)? 

In coniks-go we also AGL's implementation... We started switching to vxeddsa anyways. But currently I kept both elligator implementations: AGL's and one I wrote that aims to follow whispersystem's specs: https://github.com/coniks-sys/coniks-go/blob/vxeddsa/crypto/internal/ed25519/extra25519/extra25519.go#L396 
I didn't have the time to do more testing yet but ran against a bunch of test-vectors (and both implementations yielded the same results/hashes).

**Comment by Daeinar** [comment]
@Liamsi Unfortunately, I haven't had the time yet to check the details but yes, as soon as I know more, I let you know.

**Comment by riobard** [comment]
@Daeinar Is there any progress on this issue? I found two related issues https://github.com/agl/ed25519/pull/12 and https://github.com/Yawning/libelligator/issues/1

**Comment by nikkolasg** [comment]
Hi, thanks for the relevant links. We haven't worked on this issue so far but it is still something we plan to address in the future. Looking at Yawning's commit, it should not be too much of a hassle.

**Comment by isislovecruft** [comment]
The issue I found was separate to Yawning's issue. The code [on lines 104-107 of extra25519.go](https://github.com/agl/ed25519/blob/master/extra25519/extra25519.go#L105):

```go
var inv1 edwards25519.FieldElement
edwards25519.FeSub(&inv1, &A.Z, &A.Y)
edwards25519.FeMul(&inv1, &inv1, &A.X)
edwards25519.FeInvert(&inv1, &inv1)
```
calculates `1/X(Z-Y)` whereas the reasoning [in Adam's blogpost](https://www.imperialviolet.org/2013/12/25/elligator.html) (which seemed to check out when I worked through it) states:

> By calculating just 1/(X(Y-Z)), we can get (u,v).

Although the implementation does roundtrip, so either there is a second coordinate swap somewhere or I'm wrong (and the blogpost has a minor typo). (I haven't taken the time to hunt for a second coordinate swap, but if someone does I'd be very interested to hear about it.)

The other interesting thing I noted was that both Adam's implementation uses `sqrt(-A + 2) = sqrt(-486664)` in two places where this should have been interpreted as being `sqrt(-486660)`, which is perhaps related to Yawning's issue but I don't really care enough to re-read that horrid† paper.

† To be clear: I do not believe any of the "mistakes" in, or differences between, any of the existent elligator2 implementations to be the fault of any of the implementors. With no offence meant towards the sole author who eventually made [a nice implementation](https://sourceforge.net/projects/ed448goldilocks/), the original paper is quite frankly one of the worst written academic papers I've ever tried (dozens of times) to read, in any field. The student should have done better, the advisors should have done their actual fucking jobs, the peer reviewers should have demanded better, and the student should have been asked to provide an open-sourced reference implementation and test vectors.
